src/kem/sike/P434/AMD64/fp_x64_asm.S

```diff
@@ -0,0 +1,939 @@
+.global fpsub434_asm
+fpsub434_asm:
+  push   r12
+  push   r13
+  push   r14
+  
+  xor    rax, rax
+  mov    r8, [reg_p1]
+  mov    r9, [reg_p1+8]
+  mov    r10, [reg_p1+16]
+  mov    r11, [reg_p1+24]
+  mov    r12, [reg_p1+32]
+  mov    r13, [reg_p1+40]
+  mov    r14, [reg_p1+48]
+  sub    r8, [reg_p2] 
+  sbb    r9, [reg_p2+8] 
+  sbb    r10, [reg_p2+16] 
+  sbb    r11, [reg_p2+24] 
+  sbb    r12, [reg_p2+32] 
+  sbb    r13, [reg_p2+40] 
+  sbb    r14, [reg_p2+48]
+  sbb    rax, 0
+  
+  mov    rcx, [rip+p434x2]
+  mov    rdi, [rip+p434x2+8]
+  mov    rsi, [rip+p434x2+24]
+  and    rcx, rax
+  and    rdi, rax
+  and    rsi, rax  
+  add    r8, rcx  
+  adc    r9, rdi  
+  adc    r10, rdi  
+  adc    r11, rsi 
+  mov    [reg_p3], r8
+  mov    [reg_p3+8], r9 
+  mov    [reg_p3+16], r10 
+  mov    [reg_p3+24], r11 
+  setc   cl  
+
+  mov    r8, [rip+p434x2+32]
+  mov    rdi, [rip+p434x2+40]
+  mov    rsi, [rip+p434x2+48]
+  and    r8, rax
+  and    rdi, rax
+  and    rsi, rax  
+  bt     rcx, 0  
+  adc    r12, r8 
+  adc    r13, rdi   
+  adc    r14, rsi
+  mov    [reg_p3+32], r12 
+  mov    [reg_p3+40], r13
+  mov    [reg_p3+48], r14
+  
+  pop    r14
+  pop    r13
+  pop    r12
+  ret
+
+
+#ifdef _MULX_
+    
+///////////////////////////////////////////////////////////////// MACRO
+// Schoolbook integer multiplication, a full row at a time
+// Inputs:  memory pointers M0 and M1
+// Outputs: memory pointer C
+// Temps:   regs T0:T9
+/////////////////////////////////////////////////////////////////
+
+#ifdef _ADX_
+.macro MUL192_SCHOOL M0, M1, C, T0, T1, T2, T3, T4, T5, T6
+    mov    rdx, \M0
+    mulx   \T0, \T1, \M1     // T0:T1 = A0*B0
+    mov    \C, \T1           // C0_final
+    mulx   \T1, \T2, 8\M1    // T1:T2 = A0*B1
+    xor    rax, rax   
+    adox   \T0, \T2        
+    mulx   \T2, \T3, 16\M1   // T2:T3 = A0*B2
+    adox   \T1, \T3
+           
+    mov    rdx, 8\M0
+    mulx   \T3, \T4, \M1     // T3:T4 = A1*B0
+    adox   \T2, rax 
+    xor    rax, rax   
+    mulx   \T5, \T6, 8\M1    // T5:T6 = A1*B1
+    adox   \T4, \T0
+    mov    8\C, \T4          // C1_final  
+    adcx   \T3, \T6      
+    mulx   \T6, \T0, 16\M1   // T6:T0 = A1*B2 
+    adox   \T3, \T1  
+    adcx   \T5, \T0     
+    adcx   \T6, rax 
+    adox   \T5, \T2	
+    
+    mov    rdx, 16\M0
+    mulx   \T1, \T0, \M1     // T1:T0 = A2*B0
+    adox   \T6, rax
+    xor    rax, rax 
+    mulx   \T4, \T2, 8\M1    // T4:T2 = A2*B1
+    adox   \T0, \T3   
+    mov    16\C, \T0         // C2_final 
+    adcx   \T1, \T5    
+    mulx   \T0, \T3, 16\M1   // T0:T3 = A2*B2
+    adcx   \T4, \T6  
+    adcx   \T0, rax
+    adox   \T1, \T2
+    adox   \T3, \T4
+    adox   \T0, rax
+    mov    24\C, \T1         // C3_final
+    mov    32\C, \T3         // C4_final
+    mov    40\C, \T0         // C5_final
+.endm 
+
+.macro MUL256_SCHOOL M0, M1, C, T0, T1, T2, T3, T4, T5, T6, T7, T8, T9
+    mov    rdx, \M0
+    mulx   \T0, \T1, \M1     // T0:T1 = A0*B0
```

Comment:
isn't in going to crash if run on machine without BMI2?

Reply:
Right now these are included under a `X86_864` conditional include (e.g. https://github.com/open-quantum-safe/liboqs/blob/master/src/kem/sike/P434/P434.c#L115). Should we switch that conditional include to `HAVE_BMI2_INSTRUCTIONS`?

Reply:
I think it would be better. Just FYI I'm implementing optimized version which doesn't use BMI2 and MULX. Should be ready before end of the week. If you interested I can port it here once it's done

https://github.com/henrydcase/boringssl/tree/kris/onging/CECPQ2b_p434_aarch/third_party/sike/asm

Reply:
I've made a note on issue #512 to track fixing this.  As for adding an alternative optimized implementation, generally @christianpaquin handles the SIDH/SIKE code in liboqs on behalf of Microsoft so I'll defer to him about how to fit it in.